Introduce pl-hidden-hints element (#6334)
PrairieLearn/PrairieLearn: docs/elements.md

---

### `pl-hidden-hints` element

> Display progressive hints that reveal themselves as the number of student submissions increases for the current variant.

Reviewer: same comment about 'for the current variant'

> Hints must be opened by the user until they are revealed.

Reviewer: 'Hints must be opened by the user until they are revealed.' The meaning of this sentence is not clear to me. Are you trying to say something about when they are or aren't opened by default? It almost makes it sound like the user _must_ take some action on a hint, but they can just ignore if they want, right?

  </pl-hint>
</pl-hidden-hints>
```

#### Customizations

For the outer `pl-hidden-hints` element.

| Attribute | Type   | Default | Description                                                                                                                      |
| --------- | ------ | ------- | -------------------------------------------------------------------------------------------------------------------------------- |
| `name`    | string | `''`    | Name prefix for the key in the `params` dict. Only needed if there are separate `pl-hidden-hints` elements on the same question. |

For the inner `pl-hint` element:

| Attribute    | Type | Default | Description                                                                                              |
| ------------ | ---- | ------- | -------------------------------------------------------------------------------------------------------- |
| `show-after` | int  | -       | Number of submissions needed before the current hint is shown. If not set, hint is shown from the start. |

#### Details

Add hints to a question that are revealed with more submissions using the `show-after` attribute. By default, hints without

> `show-after` set are shown at the beginning.

Reviewer: Maybe 'are always shown' instead of 'are shown at the beginning'?

#### Example implementations

- [element/hiddenHints]

#### See also

- [`pl-question-panel` for displaying the question prompt.](#pl-question-panel-element)
- [`pl-submission-panel` for changing how a submitted answer is displayed.](#pl-submission-panel-element)

> - [`pl-answer-panel` for displaying the question's solution.](#pl-answer-panel-element)

Author: Just to emphasize that the hints can be on any panel. In the example course one of the hints is used on the submission panel.

---

### `pl-hidden-hints` element

Display progressive hints that become accessible as the number of student submissions increases for the current variant.
Hints are only open on page load when they are first revealed (when first reaching the desired submission count).
Otherwise hints start closed and must be opened by the user. The submission counter is reset when new variants are

> generated. In particular, **this element does not reveal new hints across variants.**

Reviewer 2: ```suggestion
generated. Note that **this element does not reveal new hints across variants.**
```

> generated. In particular, **this element does not reveal new hints across variants.**

Reviewer 2: I haven't fully thought this through, but I imagine that hints work best when there's a penalty for retrying; otherwise a student could just submit the same wrong answer a bunch of times to reveal all the hints. Is that how you use it in practice? Do you think it's worth mentioning that in the docs here?

Author: I think that it's worth mentioning. In our use case, the student is allowed full credit on all attempts, but there's a 1min delay between submissions. Since the hints are hidden, a student would have to sit there for a few minutes on each question to see if any were going to be revealed, and we've found this is enough of a deterrent to prevent the situation you're describing.

Author: Added some of the discussion from here and addressed all other comments @nwalters512.

#### Sample element

```html
<pl-hidden-hints>
  <pl-hint> This is a hint that will be accessible immediately. </pl-hint>

>   <pl-hint show-after="3">

Author: That's a little more explicit, will change.
